docs/design/declaring_entities.md

```diff
@@ -0,0 +1,233 @@
+        before the implementation file.
+-   A required, owning definition.
+    -   For example, `class MyClass { ... }`.
+    -   The definition might be the _only_ declaration.
+-   An optional, non-owning `extern library "<owning_library>"` declaration.
+    -   For example, `extern library "OtherLibrary" class MyClass;`.
+    -   It must be in a separate library from the definition.
+    -   The owning library's API file must import the `extern` declaration, and
+        must also contain a declaration.
+    -   The owning library's declarations must have the `extern` modifier
+        (without `library`).
+        -   For example, `extern class MyClass;`.
+
+For example, a library can have a forward declaration of an entity in the API
+file, and use the implementation file for the entity's definition. Putting the
+definition in an implementation file this way can reduce the dependencies for
+API file evaluation, improving compile time. This is commonly done with
+functions. For example:
+
+```
+library "MyLibrary";
+
+fn DoSomething();
+```
+
+```
+impl library "MyLibrary";
+
+fn DoSomething {
+  ...
+}
+```
+
+## Matching redeclarations of an entity
+
+In order to determine whether two redeclarations refer to the same entity, we
+apply the rules:
+
+-   Two named declarations _declare the same entity_ if they have the same scope
+    and the same name.
+-   One declaration _redeclares_ another if they declare the same entity and the
+    second declaration appears after the first, including the case where the
+    first declaration is imported. In this case, the second declaration is said
+    to be a _redeclaration_ of the first.
+-   Two declarations _differ_ if the sequence of tokens in the declaration
+    following the introducer keyword and the optional scope, up to the semicolon
+    or open brace, is different, except for `unused` modifiers on parameters.
+-   The program is invalid if it contains two owning declarations of the same
+    entity that differ.
+    -   The non-owned `extern library` declarations will only use semantic
+        matching for redeclarations, not syntactic matching.
+
+```carbon
+class A {
+  fn F(n: i32);
+  fn G(n: i32);
+}
+
+// ❌ Invalid: The parameter name differs.
+fn A.F(m: i32) {}
+
+// ❌ Invalid: The parameter type differs syntactically.
+fn A.G(n: (i32)) {}
+```
+
+### Details
+
+TODO: Figure out what details to pull from
+[#3762](https://github.com/carbon-language/carbon-lang/pull/3763) and
+[#3763](https://github.com/carbon-language/carbon-lang/pull/3763).
+
+## `extern` and `extern library`
+
+There are two forms of the `extern` modifier:
+
+-   On an owning declaration, `extern` limits access to the definition.
+    -   The entity must be directly imported in order to use of the definition.
+    -   An `extern library` declaration is optional.
+-   On a non-owning declaration, `extern library` provides a tool for breaking
+    dependency cycles.
```

Comment:
I believe this is also used to improve build scalability?

Reply:
Sure, added a note. More importantly, rephrased this to better capture what it actually _does_.